docs/extending/template_components.md

```diff
@@ -0,0 +1,161 @@
+
+.. attribute:: media
+
+A (possibly empty) :doc:`form media <django:topics/forms/media>` object defining JavaScript and CSS resources used by the component.
+
+.. note::
+   Any object implementing this API can be considered a valid component; it does not necessarily have to inherit from the ``Component`` class described below, and user code that works with components should not assume this (for example, it must not use ``isinstance`` to check whether a given value is a component).
+```
+
+```eval_rst
+.. _creating_template_components:
+```
+## Creating components
+
+The simplest way to create a component is to define a subclass of `wagtail.admin.ui.components.Component` and specify a `template_name` attribute on it. The rendered template will then be used as the component's HTML representation:
+
+```python
+from wagtail.admin.ui.components import Component
+
+class WelcomePanel(Component):
+    template_name = 'my_app/panels/welcome.html'
+
+
+my_welcome_panel = WelcomePanel()
+```
+
+`my_app/templates/my_app/panels/welcome.html`:
+
+```html+django
+<h1>Welcome to my app!</h1>
+```
+
+For simple cases that don't require a template, the `render_html` method can be overridden instead:
+
+```python
+from django.utils.html import format_html
+from wagtail.admin.components import Component
+
+class WelcomePanel(Component):
+    def render_html(self, parent_context):
+        return format_html("<h1>{}</h1>", "Welcome to my app!")
+```
+
+
+## Passing context to the template
+
+The `get_context_data` method can be overridden to pass context variables to the template. As with `render_html`, this receives the context dictionary from the calling template:
+
+```python
+from wagtail.admin.ui.components import Component
+
+class WelcomePanel(Component):
+    template_name = 'my_app/panels/welcome.html'
+
+    def get_context_data(self, parent_context):
+        context = super().get_context_data(parent_context)
+        context['username'] = parent_context['request'].user.username
+        return context
+```
+
+`my_app/templates/my_app/panels/welcome.html`:
+
+```html+django
+<h1>Welcome to my app, {{ username }}!</h1>
+```
+
+
+## Adding media definitions
+
+Like Django form widgets, components can specify associated JavaScript and CSS resources using either an inner `Media` class or a dynamic `media` property:
+
+```python
+class WelcomePanel(Component):
+    template_name = 'my_app/panels/welcome.html'
+
+    class Media:
+        css = {
+            'all': ('my_app/css/welcome-panel.css',)
+        }
+```
+
+
+## Using components on your own templates
```

Comment:
```suggestion
## Using components in your own templates
```

Reply:
Mmm... to me "on your own templates" feels like it's more closely pinpointing the action of putting a tag on a page - which is what this section is about - while "in your own templates" would cover the broader concepts of using components, which would include things like setting up a hook to collect components from outside code. That might just be a personal quirk though, and I'm not married to it :-)

Reply:
`your own` -> `custom`?

Reply:
This section of the docs is aimed at developers building their own areas of the admin (e.g. as third-party packages), and it feels to me like that's beyond the point where it can reasonably be called "custom".